src/util/hash.rs

```diff
@@ -25,59 +27,81 @@ use core::cmp::min;
 use hashes::Hash;
 use consensus::encode::Encodable;
 
-/// Calculates the merkle root of a list of hashes inline
-/// into the allocated slice.
+/// Calculates the merkle root of a list of *hashes*, inline (in place) in `hashes`.
 ///
 /// In most cases, you'll want to use [bitcoin_merkle_root] instead.
-pub fn bitcoin_merkle_root_inline<T>(data: &mut [T]) -> T
+///
+/// # Returns
+/// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
+/// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
+/// - `Some(merkle_root)` if length of `hashes` is greater than one.
+pub fn bitcoin_merkle_root_inline<T>(hashes: &mut [T]) -> Option<T>
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
 {
-    // Base case
-    if data.is_empty() {
-        return Default::default();
-    }
-    if data.len() < 2 {
-        return T::from_inner(data[0].into_inner());
-    }
-    // Recursion
-    for idx in 0..((data.len() + 1) / 2) {
-        let idx1 = 2 * idx;
-        let idx2 = min(idx1 + 1, data.len() - 1);
-        let mut encoder = T::engine();
-        data[idx1].consensus_encode(&mut encoder).expect("in-memory writers don't error");
-        data[idx2].consensus_encode(&mut encoder).expect("in-memory writers don't error");
-        data[idx] = T::from_engine(encoder);
+    match hashes.len() {
+        0 => None,
+        1 => Some(hashes[0]),
+        _ => Some(merkle_root_r(hashes)),
     }
-    let half_len = data.len() / 2 + data.len() % 2;
-    bitcoin_merkle_root_inline(&mut data[0..half_len])
 }
 
-/// Calculates the merkle root of an iterator of hashes.
-pub fn bitcoin_merkle_root<T, I>(mut iter: I) -> T
+/// Calculates the merkle root of an iterator of *hashes*.
+///
+/// # Returns
+/// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
+/// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
+/// - `Some(merkle_root)` if length of `hashes` is greater than one.
+pub fn bitcoin_merkle_root<T, I>(mut hashes: I) -> Option<T>
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
-          I: ExactSizeIterator<Item = T>,
+          I: Iterator<Item = T>,
 {
-    // Base case
-    if iter.len() == 0 {
-        return Default::default();
-    }
-    if iter.len() == 1 {
-        return T::from_inner(iter.next().unwrap().into_inner());
-    }
-    // Recursion
-    let half_len = iter.len() / 2 + iter.len() % 2;
-    let mut alloc = Vec::with_capacity(half_len);
-    while let Some(hash1) = iter.next() {
+    let first = hashes.next()?;
+    let second = match hashes.next() {
+        Some(second) => second,
+        None => return Some(first),
+    };
+
+    let mut hashes = iter::once(first).chain(iter::once(second)).chain(hashes);
```

Comment:
For a single-value iterator this code produces Merkle root as a hash of doubled first element and does not just returns the first element as said in the docs

Reply:
I don't see how it could possibly happen. If `hashes` returns `None` there's early return at line 63, so line 66 isn't reached. Iterator of one element is guaranteed to return `Some(_)` the first time and then it returns `None` so there's no ambiguity.

Reply:
I am saying that by returning `None` second time the second item is still converted into the copy of the first one - so the merkle root becomes not equal to the single element, but to the hash of the concatenated double first element

Reply:
It can't happen on line 66 because there's `iter::once` used and those two are guaranteed to be `Some`. On line 63 it just returns the single hash.

Since all tests are passing, perhaps you may want to write a test that causes this to fail?

Reply:
It will not happen on the line 66, it will happen below. It's quite clear from the code that if I will provide the function with just `&[hash]` it will try to compute merkle tree root as or `iter::once(hash).chain(iter::once(hash)`, resulting in an array of two, not one element.

I can work on a test case, but it will take some time.

Reply:
If you give it `&[hash]` line 66 or below will not be reached at all because the iterator will get exhausted at line 60, then at line 61 `next()` will return `None` so it will `return Some(first)` at line 63.

The test case would be useful either way and should be trivial, so anyone can do it. I'd probably do it but I need to go AFK right now.

Reply:
Oh right I see. Sorry for missing line 60